src/rules/newlineRule.ts

```diff
@@ -0,0 +1,144 @@
+/**
+* @license
+* Copyright 2017 Palantir Technologies, Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+ 
+import { getPreviousStatement, isBlock, isClassDeclaration, isFunctionWithBody } from "tsutils";
+import * as ts from "typescript";
+import * as Lint from "../index";
+
+export enum ViolationType {
+    needed = "needed",
+    needless = "needless",
+}
+
+const ALWAYS_IGNORE_OR_NEVER = {
+    enum: ["always", "never", "ignore"],
+    type: "string",
+};
+ 
+export type OptionName = "return" | "class" | "functionBlock" | "block";
+export type Option = "always" | "ignore" | "never";
+export type Options = Record<OptionName, Option>;
+
+export function parseOptions(json: Partial<Options> | undefined) {
+    const defaultValues: Options = {return: "ignore", class: "ignore", functionBlock: "ignore", block: "ignore"};
+    return {...defaultValues, ...json};
+}
+
+export class Rule extends Lint.Rules.AbstractRule {
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "newline",
+        description: "Requires or disallows blank lines before return, after class declaration, function block declaration or block.",
+        rationale: "Helps maintain a readable style in your codebase.",
+        optionsDescription: Lint.Utils.dedent`
+            Following arguments may be optionally provided:
+ 
+            * \`"return"\` checks for empty listane before return when not the only statement in the block.
```

Comment:
as discussed offline: I don't like the option names.
I'd prefer "before-return" here.
I'm also in favor of "start-of-class" to allow adding "before-class" and "after-class" later on.

And to keep naming consistent, you should prefer "function-block" over "functionBlock"